sklearn/gaussian_process/gpc.py

```diff
@@ -0,0 +1,197 @@
+
+        self.f_, _, (self.pi, self.W_sr, self.L, _, _) = \
+            self._posterior_mode(self.K_, return_temporaries=True)
+
+        return self
+
+    def predict(self, X):
+        check_is_fitted(self, ["X_fit_", "y_fit_", "K_", "f_"])
+        X = check_array(X)
+
+        # As discussed on Section 3.4.2 of GPML, for making hard binary
+        # decisions, it is enough to compute the MAP of the posterior and
+        # pass it through the link function
+        K_star = \
+            self.kernel_(self.X_fit_, X)  # K_star =k(x_star)
+        f_star = K_star.T.dot(self.y_fit_ - self.pi)  # Line 4 (Algorithm 3.2)
+
+        return f_star > 0
+
+    def predict_proba(self, X):
+        check_is_fitted(self, ["X_fit_", "y_fit_", "K_", "f_"])
+        X = check_array(X)
+
+        # Based on Algorithm 3.2 of GPML
+        K_star = self.kernel_(self.X_fit_, X)  # K_star =k(x_star)
+        f_star = K_star.T.dot(self.y_fit_ - self.pi)  # Line 4
+        v = solve(self.L, self.W_sr[:, np.newaxis] * K_star)  # Line 5
+        var_f_star = self.kernel_(X) - v.T.dot(v)  # Line 6
+
+        # Line 7:
+        # Approximate \int log(z) * N(z | f_star, var_f_star)
+        # Approximation is due to Williams & Barber, "Bayesian Classification
+        # with Gaussian Processes", Appendix A:
+        # Approximate the logistic sigmoid by a linear combination of
+        # 5 error functions
+        # See
+        # blitiri.blogspot.de/2012/11/gaussian-integral-of-error-function.html
+        # for information on how this integral can be computed
+        alpha = 1 / (2 * np.diag(var_f_star))
+        gamma = LAMBDAS * f_star
+        integrals = np.sqrt(np.pi / alpha) \
+            * erf(gamma * np.sqrt(alpha / (alpha + LAMBDAS**2))) \
+            / (2 * np.sqrt(np.diag(var_f_star) * 2 * np.pi))
+        pi_star = (COEFS * integrals).sum(axis=0) + .5 * COEFS.sum()
+
+        return pi_star
+
+    def log_marginal_likelihood(self, theta, eval_gradient=False):
+        kernel = self.kernel_.clone_with_theta(theta)
+
+        if eval_gradient:
+            K, K_gradient = kernel(self.X_fit_, eval_gradient=True)
+        else:
+            K = kernel(self.X_fit_)
+
+        K[np.diag_indices_from(K)] += self.jitter
+
+        # Return temporaries
+        f, Z, (pi, W_sr, L, b, a) = \
+            self._posterior_mode(K, return_temporaries=True)
+
+        if not eval_gradient:
+            return Z
+
+        # Compute gradient based on Algorithm 5.1 of GPML
+        d_Z = np.empty(theta.shape[0])
+        # XXX: Get rid of the np.diag() in the next line
+        R = W_sr[:, np.newaxis] * cho_solve((L, True), np.diag(W_sr))  # Line 7
+        C = solve(L, W_sr[:, np.newaxis] * K)  # Line 8
+        # Line 9:
+        s_2 = -0.5*(np.diag(K) - np.diag(C.T.dot(C))) \
+            * (pi * (1 - pi) * (1 - 2*pi))  # third derivative
+        for j in range(d_Z.shape[0]):
+            C = K_gradient[..., j]   # Line 11
```

Comment:
Do you know the dimensions of this? In that case e.g. `K_gradient[:, :, j]` is better in the sense that it will throw an error if the dimensions are wrong.